.pre-commit-config.yaml

```diff
@@ -43,7 +43,7 @@ repos:
         # supported by your project here, or alternatively use
         # pre-commit's default_language_version, see
         # https://pre-commit.com/#top_level-default_language_version
-        language_version: python3.11
+        language_version: python3.10
```

Comment:
Not sure why but pre-commit failed on circleci when set to python3.11 
 https://app.circleci.com/pipelines/github/Sceptre/sceptre/2010/workflows/e64f2159-65b5-4b30-9a1d-0ed329fefaf3/jobs/10978

Reply:
The error message:

```text
RuntimeError: failed to find interpreter for Builtin discover of python_spec='python3.11'
```

The first output of pre-commit shows that, perhaps arbitrarily, pre-commit chose python3.10 because the project doesn't support the default python3.12.0.

```text
The currently activated Python version 3.12.0 is not supported by the project (>=3.8,<3.12).
Trying to find and use a compatible version.
Using python3.10 (3.10.6)
```

Rather than change the language version of black, we should probably pull in some of changes from #1393 that this PR replaces.

* Change pyproject.toml to support Python 3.12
* Upgrade dependencies (at least PyYAML to 6.0.1. I already saw an [error like this ](https://github.com/python-poetry/poetry/issues/8287) using 6.0.0)
* Extend flake8's ignored rules (I don't know why they started flaring up either. We can address it later)

Reply:
I can reproduce the black error locally like this.

```bash
docker run -it cimg/python:3.12.0 bash -c '
git clone https://github.com/Sceptre/sceptre .
git fetch origin pull/1394/head:pr
git switch pr
sed -i "46s/language_version: python3.10/language_version: python3.11/" .pre-commit-config.yaml
poetry install --all-extras -v
poetry run pre-commit run black --all-files --show-diff-on-failure
'
```

Reply:
done.